rclpy/rclpy/context.py

```diff
@@ -28,27 +29,49 @@ def __init__(self):
         from rclpy.impl.implementation_singleton import rclpy_implementation
         self._handle = rclpy_implementation.rclpy_create_context()
         self._lock = threading.Lock()
+        self._callbacks = []
+        self._callbacks_lock = threading.Lock()
 
     @property
     def handle(self):
         return self._handle
 
     def ok(self):
+        """Check if context hasn't been shut down."""
         # imported locally to avoid loading extensions on module import
         from rclpy.impl.implementation_singleton import rclpy_implementation
         with self._lock:
             return rclpy_implementation.rclpy_ok(self._handle)
 
+    def _call_on_shutdown_callbacks(self):
+        with self._callbacks_lock:
+            for callback in self._callbacks:
+                callback()
```

Comment:
Recommend clearing `self._callbacks`. That decreases the reference count of the objects being called, and the callbacks aren't needed anymore since a context can only be shut down once.

Reply:
Alternatively, could put [`weakref.WeakMethod(callable)`](https://docs.python.org/3/library/weakref.html#weakref.WeakMethod) into the list instead of storing the callable directly.

Reply:
Addressed in https://github.com/ros2/rclpy/pull/403/commits/60c54db9e7ea98f9edb12ecd61d3c573e6df5336